Add L2 rollback

BenjaminLudwigSAP opened the pull request:
There is currently no revert logic for when SelfIP or subnet route creation/deletion fails.
This can lead to a problem when a SelfIP fails to be created (e.g. due to conflicting pool member) but the subnet route it replaces has already been deleted. In that case the deleted subnet route won't be restored.

In order to enable reverting subnet route creation/deletion, the subnet routes that existed before the L2 flow runs must be known beforehand, so that the tasks know what the desired state is. For this to work, two things are done:
- Fetching the previously existing subnet routes and SelfIPs via a dedicated flow
- Running one task for every single subnet route creation/deletion (instead of running one task to create all at once)

These changes are made in the first commit of this PR.
The second commit adds the actual revert methods.

Please read the commit messages for more detailed explanations.

BenjaminLudwigSAP committed:
Run creation/removal task for every subnet route separately

This enables us to determine existing SelfIPs/Subnet routes before the
sync flow is created, instead of figuring them out during task
execution. This is necessary to enable correct rollback of subnet
route creation/deletion (in addition to SelfIP creation/deletion),
which will be implemented in a subsequent commit.

More details:

So far, the subnet routes to be created or removed were determined in
the tasks EnsureSubnetRoutes and CleanupSubnetRoutes,
respectively. This means that their revert methods would have to
figure that out as well. However, determining the state of subnet
routes before the 'execute' method ran is not always possible, since
we don't know how many subnet routes have already been successfully
removed/created before the rollback began.

This problem is solved by determining the existing routes before
removing/creating anything, and then adding one separate task for
every subnet route deletion/creation (just like EnsureSelfIP and
RemoveSelfIP are already used for SelfIPs).

- Replace get_selfips_from_device_for_vlan with
  get_existing_selfips_and_subnet_routes_flow

- Aggregate deletion/creation of SelfIP and subnet routes in one
  superflow (created in make_sync_selfips_and_subnet_routes_flow) to
  make sure that they're rolled back together.

Other changes:

- Rename methods that return a flow to make_*_flow

- Rename cleanup flows/tasks to remove_*, since they don't determine
  anymore what's going to be removed, but that info is determined
  before the removal flow is created.

BenjaminLudwigSAP committed:
Add revert method to Ensure/Remove tasks for SelfIPs/subnet routes

Since the parent commit added existing_selfips and
existing_subnet_routes to the store, this information can now be used
to determine whether a task should be rolled back or not.

notandy commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -23,98 +23,201 @@
+
+        # find SelfIPs that are expected but not existing
+        selfips_to_create = [port for port in needed_selfips if port.id not in [p.id for p in preexisting_selfips]]
+        LOG.debug(f"{host}: SelfIPs to add for network {network.id}: {[p.id for p in selfips_to_create]}")
```
`id` of `network` is dereferenced twice, once in L172, making `network` a string and then again in L178, causing

```
ERROR octavia_f5.controller.worker.l2_sync_manager [-] Failed running sync_l2_selfips_and_subnet_routes_flow for host ...: 'str' object has no attribute 'id'
```

notandy commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -23,98 +23,201 @@
+            ensure_selfips_subflow.add(ensure_selfip_task)
+
+        # find subnet routes for subnets that need them but don't have any yet
+        subnet_route_network_part = f5_tasks.get_subnet_route_name(network.id, '')
```
`network` is a string

notandy commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -23,98 +23,201 @@
+        # make subnet route creation subflow
+        ensure_subnet_routes_subflow = unordered_flow.Flow('ensure-subnet-routes-subflow')
+        for subnet_id in subnets_to_create_routes_for:
+            subnet_route_name = f5_tasks.get_subnet_route_name(network.id, subnet_id)
```
`network` is a string

notandy commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -23,98 +23,201 @@
+        ]
+        subnets_to_create_routes_for = [s for s in subnets_that_need_routes
+                                        if s not in subnets_of_preexisting_subnet_routes]
+
```
`make_remove_selfips_and_subnet_routes_flow` has a log for subnet routes to be removed(L138), I would expect that create do the same for subnet routes to be added.

notandy commented on octavia_f5/controller/worker/tasks/f5_tasks.py:
```diff
@@ -356,58 +399,81 @@ def execute(self, network: f5_network_models.Network,
-        # prefix of subnet routes that belong to this network
-        subnet_route_network_part = get_subnet_route_name(network.id, '')
+                subnet_route, existing_subnet_routes):
+        route_name_name = subnet_route['name']
```
`subnet_route` is a string

notandy reviewed (changes requested):
Hi, so I've found multiple exceptions and stopped debugging at some point.

notandy commented:
Since this is a very large PR and it's hard to debug every corner case, I'd like to request following addition:

Add a unit tests for all (modified/added) Tasks and Flows, that
* Test Creation of Subnet routes
* Test deletion of subnet routes
* Test creation of selfips
* test removal of selfips
* test replacement of selfip with route (e.g. remove a VIP from subnet)
* test replacement of route with selfip (e.g. added a VIP to subnet)

BenjaminLudwigSAP committed:
[l2] Fix non-subscriptable variables, log subnet routes to create

BenjaminLudwigSAP committed:
[l2] Streamline names of flow functions and flow test functions

- Apply make_*_flow naming conventions to all flow-creating functions

- Name test functions after the flow they're testing.
  E.g. make_ensure_l2_flow => test_ensure_l2_flow and test_ensure_l2_flow_existing_l2

BenjaminLudwigSAP committed:
[l2][unittest] Refactor flow tests

- Add docstrings

- Move reused IDs (network/subnet) into variables

- Reformat code

BenjaminLudwigSAP committed:
[l2][unittest] Add test_ensure_selfips

BenjaminLudwigSAP committed:
[l2][unittest] Add test_ensure_subnet_route

BenjaminLudwigSAP committed:
[l2][unittest] Add test_remove_selfip

BenjaminLudwigSAP committed:
[l2] f5_tasks.RemoveSubnetRoute: Fix revert method

4f78720 fixed the passing of subnet route names to RemoveSubnetRoute.
The revert method however was not fixed in that commit. This commit
fixes the revert method. The revert method now takes the subnet route
name (instead of a subnet route dictionary) and from that constructs
the payload for restoring the subnet route, querying the subnet.

BenjaminLudwigSAP committed:
[l2] Refactor f5_tasks

- Rename some variables for more consistency

- Reorder variable assignments for more consistency

BenjaminLudwigSAP committed:
[l2][unittest] Add test_remove_subnet_route

BenjaminLudwigSAP committed:
[l2][unittest] Add test_replace_selfip_with_subnet_route

BenjaminLudwigSAP committed:
[l2][unittest] Add test_replace_subnet_route_with_selfip

BenjaminLudwigSAP committed:
[l2][unittest] Task tests: Add tests for reverting

- Add tests for reverting tasks:
  - test_revert_EnsureSelfIP
  - test_revert_EnsureSubnetRoute
  - test_revert_RemoveSelfIP
  - test_revert_RemoveSubnetRoute

- Add missing test test_EnsureSelfIP

- Rename tests for default route from *EnsureRoute* to *EnsureDefaultRoute*

BenjaminLudwigSAP committed:
[l2] Refactor f5_tasks, fix log messages

- Fix log messages in revert methods ("since it didn't exist" instead
  of "since it existed")

- Remove arguments from execute methods that are only needed by the
  revert methods

- EnsureSelfIP.execute: Adjust order of arguments for consistency with
  EnsureSubnetRoute.execute

- Rename variables (e.g. route_name to subnet_route_name)

BenjaminLudwigSAP committed:
[l2][unittest] Fix test_revert_RemoveSelfIP: SelfIPs to remove only contain ID

This is actually a bug in the code, not the test. To reveal the bug,
the test has to be fixed. This test is now running red, the code will
be fixed in a subsequent commit.

store['existing_selfips'] is a list of port objects, retrieved via
GetExistingSelfIPsForVLAN. However, these port objects contain only
the port ID. But when RemoveSelfIP is rolled back, it needs to know
more than the port ID, e.g. the subnet, which is retrieved from the
port's first fixed IP. Since the port only has an ID, the SelfIP
cannot be restored on the BigIP.

This commit fixes the test to reflect this and demonstrate the bug in
the code.

BenjaminLudwigSAP committed:
[l2][unittest] Fix: RemoveSelfIP.revert is missing critical info

When RemoveSelfIP.revert restores a deleted SelfIP on the device, it
relies on store['existing_selfips'] for the needed information.
store['existing_selfips'] is a list of port objects, retrieved via
GetExistingSelfIPsForVLAN. However, these port objects contain only
the port ID. But when RemoveSelfIP is rolled back, it needs to know
more than the port ID, e.g. the subnet, which is retrieved from the
port's first fixed IP. Since the port only has an ID, the SelfIP
cannot be restored on the BigIP.

This commit fixes GetExistingSelfIPsForVLAN to return a list of the
SelfIP dictionaries retrieved from the device, instead of a list of
port objects. This way, all the necessary information to restore a
SelfIP on the device is present.

For ease of use GetExistingSelfIPsForVLAN adds another key to the
SelfIP dictionaries: port_id. This way we only have to extract it from
the SelfIP name once.

This commit also fixes the corresponding flow/task tests to construct
store['existing_selfips'] correctly.

m-kratochvil commented:
From my end this looks good, it's a lot of code which I can't reliably review but I at least went through and checked for anything wrong like typos or mistakes in the comments but everything looks good. We had few discussions with Benjamin on Slack about the networking concepts related to routes and self-IPs on Big-IP so I trust everything is implemented correctly. But I will let Andy or Vadim press the "approve" button, from my end this PR is approved.

BenjaminLudwigSAP committed:
[l2] Fix GetExistingSelfIPsForVLAN: Address missing

velp commented:
Looks good to me. Pretty much everything is covered with unit tests (checked coverage report). The logic looks OK for me too. Let's wait final word from Andy.

notandy commented on octavia_f5/controller/worker/tasks/f5_tasks.py:
```diff
@@ -206,7 +204,7 @@ def execute(self, network: f5_network_models.Network,
             return res.json()
 
-        # Update if dict differs from on-device state
+        # Otherwise update existing subnet route (if our route isn't a subset)
```
the comment seems c&p from the subnet task? (should refer self-ips instead)

notandy reviewed (approved):
I tested a bit, and apart from minor comment typos it looks good.

BenjaminLudwigSAP committed:
Fix RemoveSubnetRoute: Always take subnet route as dictionary

RemoveSubnetRoute needs the subnet route as a dictionary (so that
revert can work), but it was supplied as a name. Fix RemoveSubnetRoute
to take a dictionary and fix usages of RemoveSubnetRoute accordingly.

More details:

GetExistingSubnetRoutesForNetwork retrieves the existing subnet
routes, returning them as list of dictionaries, taken from the device
response. This list is passed to the RemoveSubnetRoute task for it to
be able to correctly rollback. RemoveSubnetRoute is called in two
places - make_remove_selfips_and_subnet_routes_flow (for removing
UNNEEDED subnet routes) and make_remove_l2_flow (for removing ALL
subnet routes). The make_remove_l2_flow function passes the
above-mentioned list of existing subnet routes to RemoveSubnetRoute
verbatim, but make_remove_selfips_and_subnet_routes_flow erroneously
extracts the names first (via a list comprehension: [r['name'] for r
in existing_subnet_routes ...]).

- This bug was introduced in 7c0153e (which added
  GetExistingSubnetRoutesForNetwork).

- In 4f78720 I had found the erroneous call in
  make_remove_selfips_and_subnet_routes_flow and adjusted the
  RemoveSubnetRoute task to take a name instead of a dictionary, but did
  not adjust the usage in make_remove_l2_flow, thus "inverting" the
  bug. Also, I forgot to adjust the RemoveSubnetRoute.revert method,
  meaning that the execute method now worked with the call in
  make_remove_selfips_and_subnet_routes_flow but not
  make_remove_l2_flow, and the other way round for the revert method.

- In b16dbc8 I adjusted the RemoveSubnetRoute.revert method as
  well. Now both the execute and the revert methods work with the call
  in make_remove_selfips_and_subnet_routes_flow but not
  make_remove_l2_flow.

This commit fixes this error.

- Adjust RemoveSubnetRoute.execute and RemoveSubnetRoute.revert to
  take a subnet route instead of just its name. This makes calling the
  task easier, since we can fix the erroneous list comprehension (see
  above). It also makes rollback easier because the revert method now
  has all the info it needs and doesn't need to infer anything anymore.

- make_remove_l2_flow: Supply subnet_route instead of subnet_route_name

- make_remove_selfips_and_subnet_routes_flow: Supply subnet_route
  instead of subnet_route_name and - most importantly - fix list
  comprehension (to not extract the name).

- test_revert_RemoveSubnetRoute: Supply subnet_route instead of
  subnet_route_name and check that during rollback the subnet route is
  recreated with its original attributes.

BenjaminLudwigSAP committed:
Fix: Removal tasks' revert methods assume that removal worked

RemoveSelfIP.revert and RemoveSubnetRoute.revert assume that the
SelfIP/subnet route was successfully removed. So restoring can fail
and cause another exception due to HTTP 409 CONFLICT.

Check whether the SelfIP/subnet route was actually deleted before
attempting to restore.

BenjaminLudwigSAP committed:
Fix comment

m-kratochvil commented:
Checked the last three commits, look good to me, I don't see anything wrong.
